awx/api/views/__init__.py

```diff
@@ -2222,17 +2222,17 @@ class InventorySourceCredentialsList(SubListAttachDetachAPIView):
     relationship = 'credentials'
 
     def is_valid_relation(self, parent, sub, created=False):
-        # Inventory source credentials are exclusive with all other credentials
-        # subject to change for https://github.com/ansible/awx/issues/277
-        # or https://github.com/ansible/awx/issues/223
-        if parent.credentials.exists():
-            return {'msg': _("Source already has credential assigned.")}
+        # validate that this is not a duplicate of a credential already in the list
+        # same as job template credentials validation
+        if sub.unique_hash() in [cred.unique_hash() for cred in parent.credentials.all()]:
```

Comment:
The only thing I could think of that might bite us here is the special treatment of vault credentials, of which we can have more than one association provided the vault ids are different.

It looks like `unique_hash` [handles that](https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L432) for us though so I think we're ok there.

Reply:
We have a test `test_vault_credential_not_allowed_via_related` which assures that user get a 400 associating a vault credential.

While it's a conceptually coherent idea to attach a vault credential to an inventory source, it's not on the agenda, and the task logic for it is not written at all (in the current system, it would be quite complicated).

So while _this_ logic would correctly handle vault credential uniqueness, that's not yet a problem that needs to be solved.